test/org/javarosa/model/xform/SmsSerializingVisitorTest.java

```diff
@@ -0,0 +1,145 @@
+package org.javarosa.model.xform;
+
+import org.javarosa.core.PathConst;
+import org.javarosa.core.model.instance.FormInstance;
+import org.javarosa.core.services.transport.payload.ByteArrayPayload;
+import org.javarosa.core.test.FormParseInit;
+import org.javarosa.form.api.FormEntryController;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Performs a comprehensive suite of tests on the results of a form
+ * being serialized to the SMS format.
+ */
+public class SmsSerializingVisitorTest {
+    private String sms;
+    private static final String delimiter = " ";
+
+    @Before
+    public void setup() throws IOException {
+        FormParseInit formParser = new FormParseInit();
+        formParser.setFormToParse(new File(PathConst.getTestResourcePath(), "sms_form.xml").toString());
+        FormEntryController formEntryController = formParser.getFormEntryController();
+        FormInstance formInstance = formEntryController.getModel().getForm().getInstance();
+
+        SMSSerializingVisitor serializer = new SMSSerializingVisitor();
+
+        ByteArrayPayload payload = (ByteArrayPayload) serializer.createSerializedPayload(formInstance);
+
+        sms = new String(payload.getPayloadBytes(), "UTF-8").replace("\\", "").replace("\\\\", "\\");
+    }
+
+    @Test
+    public void SmsNotNull() {
+        assertNotNull(sms);
+    }
+
+    @Test
```

Comment:
Would it please you to know you could shrink this file from 145 to 82 lines, and remove most of the complexity? If you’re like me, you’ll say “Yes, yes, yes!”
Help yourself to this if you like it:
https://github.com/dcbriccetti/javarosa/commit/d52fe396f4b0e622ceb8b29d9fac5618eb4735c0

Reply:
Nice!